src/Octoshift/CodeScanningAlertService.cs

```diff
@@ -0,0 +1,209 @@
+        {
+            _log.LogInformation($"Migrating Code Scanning Analyses from '{sourceOrg}/{sourceRepo}' to '{targetOrg}/{targetRepo}'");
+
+            var analyses = await _sourceGithubApi.GetCodeScanningAnalysisForRepository(sourceOrg, sourceRepo, branch);
+            analyses = analyses.ToList();
+
+            var successCount = 0;
+            var errorCount = 0;
+
+            _log.LogVerbose($"Found {analyses.Count()} analyses to migrate.");
+
+            if (dryRun)
+            {
+                _log.LogInformation($"Running in dry-run mode. The following Sarif-Reports would now be downloaded from '{sourceOrg}/{sourceRepo}' and then uploaded to '{targetOrg}/{targetRepo}':");
+                foreach (var analysis in analyses)
+                {
+                    _log.LogInformation($"    Report of Analysis with Id '{analysis.Id}' created at {analysis.CreatedAt}.");
+                }
+                return;
+            }
+
+            foreach (var analysis in analyses)
+            {
+                var sarifReport = await _sourceGithubApi.GetSarifReport(sourceOrg, sourceRepo, analysis.Id);
+                _log.LogVerbose($"Downloaded SARIF report for analysis {analysis.Id}");
+                try
+                {
+                    await _targetGithubApi.UploadSarifReport(targetOrg, targetRepo,
+                        new SarifContainer
+                        {
+                            Sarif = sarifReport,
+                            Ref = analysis.Ref,
+                            CommitSha = analysis.CommitSha
+                        });
+                    _log.LogInformation($"Successfully Migrated report for analysis {analysis.Id}");
+                    ++successCount;
+                }
+                catch (HttpRequestException httpException)
+                {
+                    if (httpException.StatusCode.Equals(HttpStatusCode.NotFound))
+                    {
+                        _log.LogVerbose($"No commit found on target. Skipping Analysis {analysis.Id}");
+                    }
+                    else
+                    {
+                        _log.LogWarning($"Http Error {httpException.StatusCode} while migrating analysis {analysis.Id}: ${httpException.Message}");
+                    }
+                    ++errorCount;
+                }
+                catch (Exception exception)
+                {
+                    _log.LogWarning($"Fatal Error while uploading SARIF report for analysis {analysis.Id}: \n {exception.Message}");
+                    _log.LogError(exception);
+                    // Todo Maybe throw another exception here?
+                    throw;
+                }
+                _log.LogInformation($"Handled {successCount + errorCount} / {analyses.Count()} Analyses.");
+            }
+
+            _log.LogInformation($"Code Scanning Analyses done!\nSuccess-Count: {successCount}\nError-Count: {errorCount}\nOverall: {analyses.Count()}.");
```

Comment:
If some of my analyses fail to migrate, what can I do? Is it safe to run this command again?

Reply:
That is indeed a good question. Actually, as of right now, running this command multiple times is **not** safe, as the chronological order of sarif upload is important to the alert states.

E.g. if I have a repository with existing alerts from recent commits, and upload a SARIF Report that has no vulnerabilities to that repository, all alerts will be closed as `fixed` - even if the commit-sha of the SARIF Report itself is a very old one.

So, if an analysis upload fails, right now it would be most safe to find the cause, fix the issue but then remigrate the entire repository.

Let me see if I can add a function to check the target for already uploaded analysis and then just skip those - however. This will greatly increase the required API Calls.